### PR TITLE
Mooring updates

### PR DESCRIPTION
**Pull request by andermi**
It was super close, I figured it was easier to make a PR for some of the changes. Check the diffs and try running. It works!

updates for https://github.com/osrf/buoy_sim/pull/107

### CODE REVIEW

File: buoy_gazebo/src/MooringForce/MooringForce.cpp [hunk 1]
```diff
@@ -70,16 +70,16 @@ class MooringForcePrivate
...
   /// \brief N/m, weight of chain per unit length
   double w = 20.0;
 
   /// \brief radians, atan2 angle of buoy from anchor
-  double theta = 0;
+  double theta = 0.0;
 
   /// \brief Catenary equation to pass to solver
-  CatenaryHSoln catenarySoln{V, H, L};
+  std::unique_ptr<CatenaryHSoln> catenarySoln{new CatenaryHSoln(V, H, L)};
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/MooringForce.cpp:82]: H for the catenarySoln functor was not being updated

File: buoy_gazebo/src/MooringForce/MooringForce.cpp [hunk 2]
```diff
@@ -159,18 +159,20 @@ void MooringForcePrivate::UpdateVH(
...
+    (this->buoyPos[1U] - this->anchorPos[1U]) *
+    (this->buoyPos[1U] - this->anchorPos[1U]));
 
   // Update angle between buoy and anchor
-  this->theta = atan2(this->buoyPos[1] - this->anchorPos[1],
-    this->buoyPos[0] - this->anchorPos[0]);
+  this->theta = atan2(this->buoyPos[1U] - this->anchorPos[1U],
+    this->buoyPos[0U] - this->anchorPos[0U]);
+
+  this->catenarySoln.reset(new CatenaryHSoln(this->V, this->H, this->L));
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/MooringForce.cpp:175]: H for the catenarySoln functor was not being updated

File: buoy_gazebo/src/MooringForce/MooringForce.cpp [hunk 3]
```diff
@@ -243,42 +245,41 @@ void MooringForce::PreUpdate(
...
 
   Eigen::HybridNonLinearSolver<CatenaryHSoln> catenarySolver(
-    this->dataPtr->catenarySoln);
+    *this->dataPtr->catenarySoln);
   // Tolerance for error between two consecutive iterations
   catenarySolver.parameters.xtol = 0.0001;
   // Max number of calls to the function
   catenarySolver.parameters.maxfev = 1000;
+  catenarySolver.diag.setConstant(1, 1.0);
+  catenarySolver.useExternalScaling = true;  // Improves solution stability dramatically.
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/MooringForce.cpp:254]: may or may not need this now with the solver H update

File: buoy_gazebo/src/MooringForce/MooringForce.cpp [hunk 4]
```diff
@@ -243,42 +245,41 @@ void MooringForce::PreUpdate(
...
     solverInfo = catenarySolver.solveNumericalDiff(this->dataPtr->B);
 
     // Recalculate c with newly solved B
     c = CatenaryFunction::CatenaryScalingFactor(
-      this->dataPtr->V, this->dataPtr->B[0], this->dataPtr->L);
-
-    // Increment b, to iterate on B
-    b += 1.0;
+      this->dataPtr->V, this->dataPtr->B[0U], this->dataPtr->L);
   }
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/MooringForce.cpp:283]: This is the commit suggestion from the other PR

File: buoy_gazebo/src/MooringForce/MooringForce.cpp [hunk 5]
```diff
@@ -243,42 +245,41 @@ void MooringForce::PreUpdate(
...
 
   Eigen::HybridNonLinearSolver<CatenaryHSoln> catenarySolver(
-    this->dataPtr->catenarySoln);
+    *this->dataPtr->catenarySoln);
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/MooringForce.cpp:248]: now pass in dereferenced pointer

File: buoy_gazebo/src/MooringForce/CatenarySoln.hpp [hunk 6]
```diff
@@ -162,19 +164,20 @@ struct CatenaryHSoln : Functor<double>
     vSolver.parameters.xtol = 0.0001;
     // Max number of calls to the function
     vSolver.parameters.maxfev = 1000;
-
+    vSolver.diag.setConstant(1, 1.0);
+    vSolver.useExternalScaling = true;  // Improves solution stability dramatically.
```
Comment (andermi) [inline buoy_gazebo/src/MooringForce/CatenarySoln.hpp:168]: may or may not need this now with the solver H update

File: buoy_description/models/mbari_wec/model.sdf.em [hunk 7]
```diff
@@ -82,8 +82,8 @@ except NameError:
     </plugin>
 
     <plugin filename="MooringForce" name="buoy_gazebo::MooringForce">
-      <anchor_position>80 100 -84</anchor_position>
-      <enable_beyond_radius>50</enable_beyond_radius>
+      <anchor_position>80.0 100.0 -84.0</anchor_position>
+      <enable_beyond_radius>90.0</enable_beyond_radius>
```
Comment (andermi) [inline buoy_description/models/mbari_wec/model.sdf.em:86]: with V=82 and L=160, H == B when they approach 77m so I figured 90 would be a good point to turn on the mooring force

**Review by mabelzhang — APPROVED** [review]
Thank you!

**Comment by mabelzhang** [comment]
I'm going to merge this into the target PR and let the tests rerun there.

**Pull request merged**